pkg/refactor/filestore/psql/root.go

```diff
@@ -0,0 +1,122 @@
+package psql
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/direktiv/direktiv/pkg/refactor/filestore"
+	"github.com/google/uuid"
+	"gorm.io/gorm"
+)
+
+type Root struct {
+	ID uuid.UUID
+
+	Files []File `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
+
+	CreatedAt time.Time
+	UpdatedAt time.Time
+
+	db *gorm.DB
+}
+
+var _ filestore.Root = &Root{}
+
+type RootList []*Root
+
+func (r *Root) GetID() uuid.UUID {
+	return r.ID
+}
+
+func (r *Root) Delete(ctx context.Context) error {
+	res := r.db.WithContext(ctx).Delete(&Root{ID: r.ID})
+	if res.Error != nil {
+		return res.Error
+	}
+	if res.RowsAffected != 1 {
+		return fmt.Errorf("unexpedted gorm delete count, got: %d, want: %d", res.RowsAffected, 1)
+	}
+
+	return nil
+}
+
+//nolint:ireturn
+func (r *Root) CreateFile(ctx context.Context, path string, typ filestore.FileType, dataReader io.Reader) (filestore.File, error) {
+	path = filepath.Clean(path)
+	depth := strings.Count(path, "/")
+	if path == "/" {
+		depth = 0
+	}
+
+	var data []byte
+	if dataReader != nil {
+		data, _ = io.ReadAll(dataReader)
+	}
+
+	f := &File{
+		ID:     uuid.New(),
+		Path:   path,
+		Depth:  depth,
+		Data:   data,
+		Typ:    typ,
+		RootID: r.ID,
+		db:     r.db,
+	}
+	res := r.db.WithContext(ctx).Create(f)
+	if res.Error != nil {
+		return nil, res.Error
+	}
+	if res.RowsAffected != 1 {
+		return nil, fmt.Errorf("unexpedted gorm create count, got: %d, want: %d", res.RowsAffected, 1)
+	}
+
+	return f, nil
+}
+
+//nolint:ireturn
+func (r *Root) GetFile(ctx context.Context, path string) (filestore.File, error) {
+	f := &File{}
+	path = filepath.Clean(path)
+
+	res := r.db.WithContext(ctx).Where("root_id", r.ID).Where("path = ?", path).First(f)
+	if res.Error != nil {
+		return nil, res.Error
+	}
+	f.db = r.db
+
+	return f, nil
+}
+
+// nolint:ireturn
+func (r *Root) ListPath(ctx context.Context, path string) ([]filestore.File, error) {
+	var list []File
+
+	path = filepath.Clean(path)
+	depth := strings.Count(path, "/")
+	if path == "/" {
+		depth = 0
+	}
+
+	res := r.db.WithContext(ctx).
+		// Don't include file 'data' in the query. File data can be retrieved with file.GetData().
+		Select("id", "path", "depth", "root_id", "created_at", "updated_at").
+		Where("root_id", r.ID).
+		Where("depth", depth+1).
+		Where("path LIKE ?", path+"%", path).
```

Comment:
This line has a bug with a missing slash. I think it is also bugged just to look at: there's only one question mark, but we passed in two args. This is a mistake, right?